Add EIP-5192 - Minimal Soulbound NFTs (#5192)
ethereum/EIPs: EIPS/eip-5192.md

---
eip: 5192
title: Minimal Soulbound NFTs
description: Minimal standard interface for soulbinding EIP-721 NFTs
author: Tim Daubenschütz (@TimDaub)
discussions-to: https://ethereum-magicians.org/t/eip-5192-minimal-soulbound-nfts/9814
status: Review
type: Standards Track
category: ERC
created: 2022-07-01
requires: 165, 721
---

## Abstract

> This standard is an extension of [EIP-721](./eip-721.md). It proposes a minimal interface to make tokens soulbound using the feature detection functionality of [EIP-165](./eip-165.md).

Reviewer: Each EIP should be understandable with only its own text, and the EIPs listed in its `requires:` header.

I'd recommend copying over the definition of a soulbound token from your earlier EIP:

```suggestion
This standard is an extension of [EIP-721](./eip-721.md). It proposes a minimal interface to make tokens soulbound using the feature detection functionality of [EIP-165](./eip-165.md). A soulbound token is a non-fungible token bound to a single account.
```

## Motivation

The Ethereum community has expressed a need for non-transferrable, non-fungible, and socially-priced tokens similar to World of Warcraft’s soulbound items. But the lack of a token standard leads many developers to simply throw errors upon a user's invocation of transfer functionalities. Over the long term, this will lead to fragmentation and less composability.

> In this document, we outline a minimal addition to [EIP-721](./eip-721.md) that allows wallet implementers to check for a token contract's permanent (non-)transferrability using [EIP-165](./eip-165.md).

Reviewer 2: How about adding EIP-1155, the other commonly-used NFT standard? The changes you suggest should work for both without needing modification.

Author: I don't care about 1155.

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### Contract Interface

> Upon minting an EIP-5192 token, it must be permanently inseparable from the receiving account. All [EIP-721](./eip-721.md) functions of the contract that transfer the token from one account to another must throw.

Reviewer 2: ```suggestion
Upon minting a token implementing this EIP, it must be permanently inseparable from the receiving account. All [EIP-721](./eip-721.md) functions of the contract that transfer the token from one account to another must throw.
```

> To aid recognition that an [EIP-721](./eip-721.md) token implements soulbinding via `EIP-5192`, upon calling [EIP-721](./eip-721.md)'s `function supportsInterface(bytes4 interfaceID) external view returns (bool)` with EIP-5192's `SOULBOUND_VALUE=0x9e7d7f8` as `bytes4 interfaceID`, a contract implementing `EIP-5192` must return `true`.

Reviewer 2: I think perhaps something even simpler (and much more versatile!) is to add a `function soulbound(uint256 tokenId) external view returns (bool tokenIdSoulbound)`, which would even allow for soulbounding to be managed on a per-tokenId basis. This is by no means necessary though.

Reviewer 3: This can be known implicitly by calling `ownerOf(tokenId)` in the [4973 spec](https://eips.ethereum.org/EIPS/eip-4973)

Author: > I like the interface you proposed, a token's "soulboundness" can be known implicitly by calling `ownerOf(tokenId)` in the [4973 spec](https://eips.ethereum.org/EIPS/eip-4973).

no. Unless I'm missing something

> To aid recognition that an [EIP-721](./eip-721.md) token implements soulbinding via `EIP-5192`, upon calling [EIP-721](./eip-721.md)'s `function supportsInterface(bytes4 interfaceID) external view returns (bool)` with EIP-5192's `SOULBOUND_VALUE=0x9e7d7f8` as `bytes4 interfaceID`, a contract implementing `EIP-5192` must return `true`.

Reviewer 2: Here is something that _does_ need to be changed:
```suggestion
To aid recognition that an [EIP-721](./eip-721.md) token implements soulbinding via this EIP upon calling [EIP-721](./eip-721.md)'s `function supportsInterface(bytes4 interfaceID) external view returns (bool)` with EIP-5192's `SOULBOUND_VALUE=0x9e7d7f8` as `bytes4 interfaceID`, a contract implementing this EIP must return `true`.
```

> To aid recognition that an [EIP-721](./eip-721.md) token implements soulbinding via `EIP-5192`, upon calling [EIP-721](./eip-721.md)'s `function supportsInterface(bytes4 interfaceID) external view returns (bool)` with EIP-5192's `SOULBOUND_VALUE=0x9e7d7f8` as `bytes4 interfaceID`, a contract implementing `EIP-5192` must return `true`.

Reviewer: ```suggestion
To aid recognition that an [EIP-721](./eip-721.md) token implements soulbinding via this proposal, upon calling [EIP-721](./eip-721.md)'s `function supportsInterface(bytes4 interfaceID) external view returns (bool)` with `SOULBOUND_VALUE=0x9e7d7f8` as `bytes4 interfaceID`, a contract implementing this proposal must return `true`.
```

> We compute `SOULBOUND_VALUE` by running the string "soulbound" through `keccak256` and casting it to a `bytes4`:

Reviewer 2: Nit: This _technically_ doesn't follow EIP-165. According to EIP-165, it should be the identifier of an interface. This is, as you have described, not an identifier of an interface.

Author: fixed in https://github.com/ethereum/EIPs/pull/5192/commits/f6b0b5740678249cdd9a0fb73fa834f76e1d86a2

```solidity
bytes4 constant SOULBOUND_VALUE = bytes4(keccak256("soulbound")); // 0x9e7ed7f8;
```

## Rationale

The above model is the simplest possible path towards a canonical interface for Soulbound tokens. It reflects upon the numerous Soulbound token implementations that simply revert upon transfers.

## Backwards Compatibility

> `EIP-5192` is fully backward compatible with [EIP-721](./eip-721.md).

Reviewer: ```suggestion
This proposal is fully backward compatible with [EIP-721](./eip-721.md).
```

The Ethereum community has expressed a need for non-transferrable, non-fungible, and socially-priced tokens similar to World of Warcraft’s soulbound items. But the lack of a token standard leads many developers to simply throw errors upon a user's invocation of transfer functionalities. Over the long term, this will lead to fragmentation and less composability.

In this document, we outline a minimal addition to [EIP-721](./eip-721.md) that allows wallet implementers to check for a token contract's permanent (non-)transferrability using [EIP-165](./eip-165.md).

## Specification

The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY" and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### Contract Interface

Upon minting a token implementing this EIP, it must be permanently inseparable from the receiving account. All [EIP-721](./eip-721.md) functions of the contract that transfer the token from one account to another must throw.

```solidity
// SPDX-License-Identifier: CC0-1.0
pragma solidity ^0.8.0;

interface IERC5192 {
  /// @notice Returns the locking status of an Soulbound Token

>   /// @dev SBTs assigned to zero address are considered invalid, and querie

Reviewer 4: ```suggestion
  /// @dev SBTs assigned to zero address are considered invalid, and queries
```
